python/ray/tune/pbt.py

```diff
@@ -0,0 +1,174 @@
+        grace_period (str): Period of time, in which algorithm will not compare 
+            model to other models
+        ready (func): Determines in Trial is ready to undergo the comparison,
+            exploitation, and exploration process. Should return boolean
+        hyperparamter_mutations (dict); Possible values that each hyperparameter can mutate
+            to, as certain hyperparameters only work with certain values
+    """
+    
+    def __init__(
+            self, time_attr="training_iteration", reward_attr='episode_reward_mean',
+            grace_period=4.0, ready=None, hyperparameter_mutations=None):
+        FIFOScheduler.__init__(self)
+        self._stopped_trials = set()
+        self._completed_trials = set()
+        self._results = collections.defaultdict(list)
+        self._time_tracker = {}
+        self._grace_period = grace_period
+        self._reward_attr = reward_attr
+        self._time_attr = time_attr
+        self._ready_func = ready
+        self._hyperparameter_mutations = hyperparameter_mutations
+
+    def on_trial_result(self, trial_runner, trial, result):
+
+        print("wowee, trial had a result")
+        print(getattr(trial, "config"))
+        self._results[trial].append(result)
+        time = getattr(result, self._time_attr)
+        ready = False
+        #check model is ready to undergo mutation, based on user function or default
+        #function
+        if time > self._grace_period:
+            ready = self._ready_func(result, trial) if self._ready_func is not None else self._default_ready(result, trial, time)
+        else:
+            print("grace period")
+            ready = False
+        if ready:
+            print("ready to undergo mutation")
+            print ("----")
+            print("Current Trial is: {0}".format(trial))
+            #get best trial for current time
+            best_trial = self._get_best_trial(result, time)
+            print("Best Trial is: {0}".format(best_trial))
+            print(best_trial.config)
+            
+            #if current trial is the best trial (as in same hyperparameters), do nothing
+            if trial.config == best_trial.config:
+                print("current trial is best trial")
+                return TrialScheduler.CONTINUE
+            else:
+                #setup the trial runner in case it
+                #create a new trial, with the same weights and hyperparamters as the best_trial
+                mutate_string = "_mutated@" + str(time)
+                config_dictionary = copy.deepcopy(best_trial.config)
+                #if trial has finished/paused, restart the runner, so that checkpoint works
+                best_trial_was_stopped = False
+                if best_trial.status == "TERMINATED" or best_trial.status == "PAUSED":
+                    best_trial_was_stopped = True
+                    best_trial._setup_runner()
+                
+                if self._hyperparameter_mutations != None:
```

Comment:
Let's move this into a standalone `exploit(hyperparams, hyperparam_mutations)` function, that way you can easily unit test the mutation function in isolation.